src/jupyterlabcontroller/services/source/base.py

```diff
@@ -0,0 +1,135 @@
+"""Base class for image sources."""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from typing import Mapping
+
+from structlog.stdlib import BoundLogger
+
+from ...models.domain.docker import DockerReference
+from ...models.domain.form import MenuImage
+from ...models.domain.kubernetes import KubernetesNodeImage
+from ...models.domain.rspimage import RSPImage, RSPImageCollection
+from ...models.v1.prepuller import PrepulledImage
+from ...models.v1.prepuller_config import PrepullerConfig
+
+__all__ = ["ImageSource"]
+
+
+class ImageSource(metaclass=ABCMeta):
```

Comment:
Neat.  I like the idea of an abstract superclass that these things inherit from.  That will also make it easy if we determine that ghcr has a better-than-Docker API or something.